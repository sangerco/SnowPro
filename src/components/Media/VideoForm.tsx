import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { NewVideoData, createVideo } from "../../redux/slices/mediaSlices";
import { fetchAllTags } from "../../redux/slices/tagSlice";
import {
  Form,
  Button,
  Dropdown,
  DropdownItemProps,
  Container,
  Divider,
} from "semantic-ui-react";
import TagForm from "../Tags/TagForm";

const VideoForm: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const dispatch = useDispatch<AppDispatch>();

  useEffect(() => {
    dispatch(fetchAllTags());
  }, [dispatch]);

  const tags = useSelector((state: RootState) => state.tags);

  let tagOptions:
    | { key: string; text: string; value: string }[]
    | DropdownItemProps[]
    | undefined;

  if (tags.tags) {
    tagOptions = tags.tags.map((tag) => ({
      key: tag.tagId,
      text: tag.tag,
      value: tag.tagId,
    }));
  } else {
    tagOptions = [];
  }

  const initialState: NewVideoData = {
    userId: userId ?? "",
    link: "",
    about: "",
    tagIds: [],
  };

  const [formData, setFormData] = useState(initialState);
  const [showCreateNewTagForm, setShowCreateNewTagForm] = useState(false);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(createVideo(formData));

    setFormData(initialState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  return (
    <Container fluid>
      <Form onSubmit={handleSubmit}>
        <Form.Field>
          <label>Photo Link</label>
          <input type="text" onChange={handleChange} value={formData.link} />
        </Form.Field>
        <Form.Field>
          <label>About</label>
          <input type="text" onChange={handleChange} value={formData.about} />
        </Form.Field>
        <Divider />
        <label>Tags</label>
        <Dropdown
          clearable
          options={tagOptions}
          multiple
          fluid
          selection
          value={formData.tagIds}
        />
        <Button onClick={() => setShowCreateNewTagForm(true)} size="small">
          Create New Tag?
        </Button>
        {showCreateNewTagForm && <TagForm />}
        <Divider />
      </Form>
    </Container>
  );
};

export default VideoForm;
