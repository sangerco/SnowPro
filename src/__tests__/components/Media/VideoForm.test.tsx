/* eslint-disable testing-library/no-node-access */
import React from "react";
import VideoForm from "../../../components/Media/VideoForm";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import nock from "nock";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Video Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    media: {
      photo: null,
      photos: [],
      video: null,
      videos: [],
      media: null,
      loading: false,
      error: null,
    },
  };
  const store = mockStore(initialState);
  it("renders with crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <VideoForm />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <VideoForm />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render a form", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <VideoForm />
        </BrowserRouter>
      </Provider>
    );
    const videoLink = screen.getByLabelText(
      "Video Link - must be YouTube link"
    );
    const about = screen.getByLabelText("About");
    const videoForm = screen.getByTestId("video-form");
    expect(videoLink).toBeInTheDocument();
    expect(about).toBeInTheDocument();
    expect(videoForm).toBeInTheDocument();
  });

  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/videos")
      .reply(201, {
        video: {
          id: "video-1",
          userId: "testuser",
          link: "http://www.testlinks.com/video1.mp4",
          about: "test video about",
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <VideoForm />
        </BrowserRouter>
      </Provider>
    );
    expect(nock.pendingMocks().length).toBe(1);

    const videoLink = screen.getByLabelText(
      "Video Link - must be YouTube link"
    );
    const videoAbout = screen.getByLabelText("About");

    fireEvent.change(videoLink, {
      target: { value: "https://www.youtube.com/testvideo" },
    });
    fireEvent.change(videoAbout, { target: { value: "test video about" } });

    const videoForm = await screen.findByTestId("video-form");
    fireEvent.submit(videoForm);

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
