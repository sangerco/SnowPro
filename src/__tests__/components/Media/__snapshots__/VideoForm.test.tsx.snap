// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Video Form matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui fluid container"
    style="margin: 10px;"
  >
    <form
      class="ui form"
      data-testid="video-form"
    >
      <div
        class="field"
      >
        <label
          for="videoLink"
        >
          Video Link - must be YouTube link
        </label>
        <input
          id="videoLink"
          name="link"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="about"
        >
          About
        </label>
        <input
          id="about"
          name="about"
          type="text"
          value=""
        />
      </div>
      <div
        class="ui divider"
      />
      <button
        class="ui green small button"
        type="submit"
      >
        Send it.
      </button>
    </form>
  </div>
</DocumentFragment>
`;
